sklearn/feature_selection/from_model.py

```diff
@@ -123,25 +129,52 @@ class SelectFromModel(BaseEstimator, SelectorMixin, MetaEstimatorMixin):
     threshold_ : float
         The threshold value used for feature selection.
     """
-    def __init__(self, estimator, threshold=None, prefit=False, norm_order=1):
+    def __init__(self, estimator, threshold=None, prefit=False,
+                 max_features=None, norm_order=1):
         self.estimator = estimator
         self.threshold = threshold
         self.prefit = prefit
+        self.max_features = max_features
         self.norm_order = norm_order
 
+    def _check_max_features(self, X, max_features):
+        if self.max_features is None or self.max_features == 'all':
+            return
+
+        if isinstance(self.max_features, numbers.Integral):
+            if 0 <= self.max_features <= X.shape[1]:
+                return
+
+        raise ValueError(
+            "max_features should be >=0, <= n_features or 'all';"
+            " got %r. Use max_features='all' to return all features."
+            % self.max_features)
+
+    def _check_params(self, X, y):
+        X, y = check_X_y(X, y)
+        self._check_max_features(X, self.max_features)
+
     def _get_support_mask(self):
         # SelectFromModel can directly call on transform.
         if self.prefit:
             estimator = self.estimator
         elif hasattr(self, 'estimator_'):
             estimator = self.estimator_
         else:
-            raise ValueError(
-                'Either fit SelectFromModel before transform or set "prefit='
-                'True" and pass a fitted estimator to the constructor.')
+            raise ValueError('Either fit the model before transform or set'
+                             ' "prefit=True" while passing the fitted'
+                             ' estimator to the constructor.')
         scores = _get_feature_importances(estimator, self.norm_order)
         threshold = _calculate_threshold(estimator, scores, self.threshold)
-        return scores >= threshold
+        mask = np.zeros_like(scores, dtype=bool)
+        n_features_to_select = self.max_features
+        if self.max_features == 'all':
+            n_features_to_select = scores.size
+        candidate_indices = np.argsort(-scores,
```

Comment:
This sort is unnecessary in the default `max_features` case, and so seems to be wasted computation.

Note that an alternative way to implement this, in O(n) time, is to just set `threshold=max(threshold, np.percentile(scores, 100 * max_features / n_features))`, and then handle ties explicitly if too many features are selected. I'm happy with mergesort for readability and comparison to SelectKBest

Reply:
For now, I just added a check to avoid sorting in the default case. I think the tiebreaking code required for the percentile approach would decrease readability significantly (as you mention). Of course, if performance becomes an issue, I can always go back and change it.